diff -uNr orange-pi-5.4-bk/arch/arm64/boot/dts/allwinner/sun50i-a64-sopine-baseboard.dts orange-pi-5.4/arch/arm64/boot/dts/allwinner/sun50i-a64-sopine-baseboard.dts
--- orange-pi-5.4-bk/arch/arm64/boot/dts/allwinner/sun50i-a64-sopine-baseboard.dts	2020-02-29 13:33:14.960781306 +0100
+++ orange-pi-5.4/arch/arm64/boot/dts/allwinner/sun50i-a64-sopine-baseboard.dts	2020-02-29 13:48:42.224214185 +0100
@@ -46,7 +46,7 @@
 /dts-v1/;
 
 #include "sun50i-a64-sopine.dtsi"
-
+#include <dt-bindings/pwm/pwm.h>
 / {
 	model = "SoPine with baseboard";
 	compatible = "pine64,sopine-baseboard", "pine64,sopine",
@@ -61,6 +61,14 @@
 		stdout-path = "serial0:115200n8";
 	};
 
+	backlight: backlight {
+		compatible = "pwm-backlight";
+		pwms = <&r_pwm 0 50000 PWM_POLARITY_INVERTED>;
+		brightness-levels = <1 2 4 8 16 32 64 128 512>;
+		default-brightness-level = <8>;
+		enable-gpios = <&pio 7 10 GPIO_ACTIVE_HIGH>; /* LCD-BL-EN: PH10 */
+	};
+
 	hdmi-connector {
 		compatible = "hdmi-connector";
 		type = "a";
@@ -104,6 +112,24 @@
 	status = "okay";
 };
 
+&dphy {
+	status = "okay";
+};
+
+&dsi {
+	vcc-dsi-supply = <&reg_dldo1>;		/* VCC3V3-DSI */
+	status = "okay";
+
+	panel@0 {
+		compatible = "feiyang,fy07024di26a30d";
+		reg = <0>;
+		avdd-supply = <&reg_dc1sw>;	/* VCC-LCD */
+		dvdd-supply = <&reg_dldo2>;	/* VCC-MIPI */
+		reset-gpios = <&pio 3 24 GPIO_ACTIVE_HIGH>; /* LCD-RST: PD24 */
+		backlight = <&backlight>;
+	};
+};
+
 &ehci0 {
 	status = "okay";
 };
@@ -248,6 +274,10 @@
 	vcc-hdmi-supply = <&reg_dldo1>;
 };
 
+&r_pwm {
+	status = "okay";
+};
+
 &sound {
 	simple-audio-card,aux-devs = <&codec_analog>;
 	simple-audio-card,widgets = "Microphone", "Microphone Jack",
